### PR TITLE
Typescript support (entry level only)

### PR DESCRIPTION
**Pull request by alienkarma**
Attempt to fix issue #41 [Enhance Public AgenticJS API TypeScript](#41)

- Added types for relevant entry API
- Missing type for unknown context (WIP)
- Added "test:types" script for testing typescript validation
- Added package type: module

### CODE REVIEW

File: types/index.d.ts [hunk 1]
```diff
@@ -1,45 +1,98 @@
...
-}
+  export class Task {
+    id: string;
+    title: string;
+    description: string;
+    expectedOutput: string;
+    isDeliverable: boolean;
+    agent: BaseAgent;
+    status: TASK_STATUS_enum;
+    result: any; // ? Need more context
```
Comment (darielnoel) [inline types/index.d.ts:64]: It is the result of a given task. Once an Agent has finished working on it. It is mean to be a read only attribute.

File: types/index.d.ts [hunk 2]
```diff
@@ -1,45 +1,98 @@
...
-        Object.assign(this, props);
-    }    
+    constructor(params: ITeamParams);
 
-    // Implementation details will be defined later... 
-    // We're scheduling a meeting to plan the meeting about this.
+    start(): Promise<void>;
+    getStore(): TStore;
+    useStore(): TStore;
+    subscribeToChanges(
```
Comment (darielnoel) [inline types/index.d.ts:93]: The listener is a callback function that gets executed when the specified properties change.

Example:

```js
subscribeToChanges((newValues) => {
    console.log('State updated:', newValues);
}, ['user', 'isLoggedIn']);

```

**Comment by darielnoel** [comment]
Hi `@alienkarma`,

Thank you so much for your contributions! Adding TypeScript to AgenticJS is a significant enhancement that will undoubtedly benefit many developers. 

I’ve reviewed your pull request and have a few suggestions. Your expertise is truly valuable to us, and your efforts are greatly appreciated. Please know that these suggestions are meant to build on the fantastic foundation you’ve already laid, without any pressure.

1. **Link to Issue & Description in PR**: It would be great if you could include a link to the related issue and a brief description in your PR to provide context for other collaborators.

2. **JSDoc Annotations**: Do you think it would be beneficial to add JSDoc annotations to the type definitions? I’m under the impression that doing so could enhance IDE autocompletion features for TypeScript users, making the library more intuitive to use. What's your take on this? Would this addition be helpful in your view? If so you can get it from here(https://docs.agenticjs.com/category/core-concepts) 

3. **TypeScript Example Project**: When you have a moment, could you create a small TypeScript example in the [Playground](https://github.com/AI-Champions/AgenticJS/tree/main/playground)? Including a README with usage examples would be wonderful, and we’d love to showcase this in our main documentation later on.

Thank you again for your hard work and dedication. I’m here to support you in any way I can.

**Comment by alienkarma** [comment]
Nice.
Thank you for the feedback.
I will implement the necessary changes and update once done.

File: types/index.d.ts (hunk 2 above)
Comment (alienkarma) [reply types/index.d.ts:93]: This example has helped clarify and solve the type here.

File: types/index.d.ts (hunk 1 above)
Comment (alienkarma) [reply types/index.d.ts:64]: Can give an example or reference to what the type/structure of the response would be like?

**Comment by alienkarma** [comment]
Hi @darielnoel 
I added the JSDocs and also created a typescript version of the nodejs playground example.
The type seems to be working (fixed a few type issues because of this as well).
I will work on a proper typescript react example later on.

To add the types it has to be declared as a separate package `@types/agenticjs`, assuming the module name is `agenticjs`.
I am attaching the node_modules directory for more clarity. (The index.js and index.js.map are module.esm.js and module.esm.js.map files respectively)

image

I am not sure yet how to produce a proper distributable output yet using rollup. (except as an external type package)
Anyways, let me know if you have any more feedback and/or changes.

**Comment by darielnoel** [comment]
Thank you @alienkarma, I'm sorry for the back and forth. Just trying to make sure that every feature we add has a solid foundation. Here's some of the philosophy we are trying to pursue.

https://resend.com/philosophy

Thanks for your patience and understanding 🤗

**Comment by pelikhan** [comment]
Very much looking for TypeScript friendlyness; it really helps diving into a new codebase.

**Comment by alienkarma** [comment]
Hi @darielnoel
I added a typed package builder that we can import and use as a package for testing (and maybe distribution).
I have also updated the nodejs-ts playground example using the built package and also updated the readme for it as well.

As always, let me know if any changes/feedback. (I really appreciate it 😄)

File: types/index.d.ts (hunk 1 above)
Comment (darielnoel) [reply types/index.d.ts]: A string.

**Pull request merged**